File: StudyBuddy.BackendAngular/src/app/model/challenge.ts
import { Helper } from "../shared/helper";

export class Challenge {
    id:string = "";
    name:string = "";;
    description:string = "";;
    points:number = 1;
    validity_start:string;
    validity_end:string;
    category:number = 1;
    owner:string = "";;
    created:string;
    prove:number = 1;
    parent:string = "";;
    study_program:string = "";
    enrolled_since:string = "";

    constructor() { 
        let today = (new Date()).toISOString().split('T')[0];
        this.validity_start = today;
        this.validity_end = today;
        this.created = today;
    }

    copyValues(values) {
        this.name = values.name;
        this.description = values.description;
        this.points = +values.points;
        this.validity_start = values.validity_start;
        this.validity_end = values.validity_end;
        this.category = +values.category;
        this.prove = +values.prove;
        this.study_program = values.study_program;
        this.enrolled_since = values.enrolled_since;
    }

    validityStartAsDate():Date {
        return Helper.convertToDate(this.validity_start);
    }

    validityEndAsDate():Date {
        return Helper.convertToDate(this.validity_end);
    }

    isPeriodValid():boolean {
        return this.validityEndAsDate() >= this.validityStartAsDate();
    }

    isInPeriod(d:Date):boolean {
        return d >= this.validityStartAsDate() && d <= this.validityEndAsDate();
    }
}

File: StudyBuddy.Model/Challenge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Serialization;

namespace StudyBuddy.Model
{
    public class Challenge : Entity
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public int Points { get; set; } = 1;
        public DateTime ValidityStart { get; set; } = DateTime.Now.Date;
        public DateTime ValidityEnd { get; set; } = DateTime.Now.Date;
        public ChallengeCategory Category { get; set; } = ChallengeCategory.Learning;
        public int OwnerID { get; set; }
        public DateTime Created { get; set; } = DateTime.Now.Date;
        public ChallengeProve Prove { get; set; } = ChallengeProve.ByTrust;
        public int? SeriesParentID { get; set; }
        public string Tags { get; set; }

        [JsonIgnore]
        public string ValidityText
        {
            get
            {
                var result = "";

                if (ValidityStart.Year == ValidityEnd.Year && ValidityStart.Month == ValidityEnd.Month)
                    result = ValidityStart.ToString("dd.") + "-" + ValidityEnd.ToString("dd.MM.yyyy");
                else if (ValidityStart.Date == ValidityEnd.Date)
                    result = ValidityEnd.ToString("dd.MM.yyyy");
                else
                    result = ValidityStart.ToString("dd.MM.yyyy") + "-" + ValidityEnd.ToString("dd.MM.yyyy");

                return result;
            }
        }

        [JsonIgnore]
        public string PointsText
        {
            get
            {
                return Points > 1 ? Points + " Punkte" : Points + " Punkt";
            }
        }

        [JsonIgnore]
        public char CategorySymbol
        {
            get
            {
                // GraduationCap
                if (Category == ChallengeCategory.Learning)
                    return '\uf19d';

                // PeopleArrows
                if (Category == ChallengeCategory.Networking)
                    return '\ue068';

                return '\uf0ae';
            }
        }

        public Challenge Clone()
        {
            var clone = new Challenge();
            clone.Name = Name;
            clone.Description = Description;
            clone.Points = Points;
            clone.ValidityStart = ValidityStart;
            clone.ValidityEnd = ValidityEnd;
            clone.Category = Category;
            clone.OwnerID = OwnerID;
            clone.Created = DateTime.Now.Date;
            clone.Prove = Prove;
            clone.SeriesParentID = SeriesParentID;
            return clone;
        }

        private IEnumerable<string> SplitSearchText(string search_text)
        {
            return search_text.ToLower().Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
        }

        private bool Contains(string property, IEnumerable<string> keywords)
        {
            return property == null ? false : keywords.All(kw => property.ToLower().Contains(kw));
        }

        public bool Contains(string search_text)
        {
            var keywords = SplitSearchText(search_text);
            return Contains(Name + Description + Tags, keywords);
        }
    }
}